Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-40.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: c2a49a91d3aed157d1e8b9795eb330821e35079b
Containerized:False
Number of Error(s): 4
src-react/components/forms/Banner.tsx(52,26): error TS2339: Property updateBanner does not exist on type Activity.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:337: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:12844: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:182: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-11-12_19-19-29
time logs started
Prep time   :    16 sec (00:00:16)
Build time  :   157 sec (00:02:37)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   173 sec (00:02:53)
time logs end

[Build error:](https://gist.github.com/b2f2122b1c96284ab80cfdca08b92c8d)
Build log [Build log](https://gist.github.com/HPCCSmoketest/b2f2122b1c96284ab80cfdca08b92c8d#file-result-21-11-12-19-16-22-log)

--- a/esp/src/package.json
+++ b/esp/src/package.json
@@ -42,7 +42,7 @@
     "@hpcc-js/chart": "2.69.0",
     "@hpcc-js/codemirror": "2.52.0",
     "@hpcc-js/common": "2.59.0",
-    "@hpcc-js/comms": "2.63.0",
+    "@hpcc-js/comms": "2.64.0",
     "@hpcc-js/dataflow": "3.0.1",
     "@hpcc-js/eclwatch": "2.63.0",
     "@hpcc-js/graph": "2.71.0",

--- a/esp/src/src-react/components/Title.tsx
+++ b/esp/src/src-react/components/Title.tsx
@@ -8,6 +8,7 @@ import * as WsAccount from "src/ws_account";
 import * as cookie from "dojo/cookie";
 
 import nlsHPCC from "src/nlsHPCC";
+import { useBanner } from "../hooks/banner";
 import { useECLWatchLogger } from "../hooks/logging";
 import { useGlobalStore } from "../hooks/store";
 import * as Utility from "src/Utility";
@@ -42,6 +43,9 @@ export const DevTitle: React.FunctionComponent<DevTitleProps> = ({
     const [environmentTitle, setEnvironmentTitle] = useGlobalStore("HPCCPlatformWidget_Toolbar_Text", toolbarThemeDefaults.text);
     const [titlebarColor, setTitlebarColor] = useGlobalStore("HPCCPlatformWidget_Toolbar_Color", toolbarThemeDefaults.color);
 
+    const [showBannerConfig, setShowBannerConfig] = React.useState(false);
+    const [BannerMessageBar, BannerConfig] = useBanner({ showForm: showBannerConfig, setShowForm: setShowBannerConfig });
+
     const personaProps: IPersonaSharedProps = React.useMemo(() => {
         return {
             text: currentUser?.firstName + " " + currentUser?.lastName,
@@ -68,7 +72,7 @@ export const DevTitle: React.FunctionComponent<DevTitleProps> = ({
             items: [
                 { key: "legacy", text: nlsHPCC.OpenLegacyECLWatch, href: "/esp/files/stub.htm" },
                 { key: "divider_0", itemType: ContextualMenuItemType.Divider },
-                { key: "banner", text: nlsHPCC.SetBanner },
+                { key: "banner", text: nlsHPCC.SetBanner, onClick: () => setShowBannerConfig(true) },
                 { key: "toolbar", text: nlsHPCC.SetToolbar, onClick: () => setShowTitlebarConfig(true) },
                 { key: "divider_1", itemType: ContextualMenuItemType.Divider },
                 { key: "docs", href: "https://hpccsystems.com/training/documentation/", text: nlsHPCC.Documentation, target: "_blank" },
@@ -143,6 +147,7 @@ export const DevTitle: React.FunctionComponent<DevTitleProps> = ({
     }, [environmentTitle]);
 
     return <div style={{ backgroundColor: titlebarColor ?? theme.palette.themeLight }}>
+        <BannerMessageBar />
         <Stack horizontal verticalAlign="center" horizontalAlign="space-between">
             <Stack.Item align="center">
                 <Stack horizontal>
@@ -200,6 +205,7 @@ export const DevTitle: React.FunctionComponent<DevTitleProps> = ({
             environmentTitle={environmentTitle} setEnvironmentTitle={setEnvironmentTitle}
             titlebarColor={titlebarColor} setTitlebarColor={setTitlebarColor}
         />
+        <BannerConfig />
     </div>;
 };
 

--- a/esp/src/src-react/hooks/activity.ts
+++ b/esp/src/src-react/hooks/activity.ts
@@ -1,0 +1,30 @@
+import * as React from "react";
+import { Activity } from "@hpcc-js/comms";
+import { useCounter } from "./workunit";
+
+export function useActivity(): [Activity, number, () => void] {
+
+    const [activity, setActivity] = React.useState<Activity>();
+    const [lastUpdate, setLastUpdate] = React.useState(Date.now());
+    const [count, increment] = useCounter();
+
+    React.useEffect(() => {
+        const activity = Activity.attach({ baseUrl: "" });
+        let active = true;
+        let handle;
+        activity.lazyRefresh().then(() => {
+            if (active) {
+                setActivity(activity);
+                handle = activity.watch(() => {
+                    setLastUpdate(Date.now());
+                });
+            }
+        });
+        return () => {
+            active = false;
+            handle.release();
+        }
+    }, [count]);
+
+    return [activity, lastUpdate, increment];
+}

--- a/esp/src/src-react/hooks/banner.tsx
+++ b/esp/src/src-react/hooks/banner.tsx
@@ -1,0 +1,194 @@
+import * as React from "react";
+import { Checkbox, ColorPicker, DefaultButton, getColorFromString, IColor, Label, MessageBar, MessageBarType, PrimaryButton, Stack, TextField } from "@fluentui/react";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { useActivity } from "./activity";
+import { MessageBox } from "../layouts/MessageBox";
+
+interface BannerConfigValues {
+    BannerAction: number;
+    BannerContent: string;
+    BannerColor: string;
+    BannerSize: string;
+    BannerScroll: string;
+}
+
+const defaultValues: BannerConfigValues = {
+    BannerAction: 0,
+    BannerContent: "",
+    BannerColor: "",
+    BannerSize: "",
+    BannerScroll: ""
+};
+
+interface useBannerProps {
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+const white = getColorFromString("#ffffff");
+
+export function useBanner({ showForm, setShowForm }: useBannerProps): [React.FunctionComponent, React.FunctionComponent] {
+
+    const [activity] = useActivity();
+
+    const [bannerColor, setBannerColor] = React.useState(activity?.BannerColor || "black");
+    const [bannerMessage, setBannerMessage] = React.useState(activity?.BannerContent || "");
+    const [bannerSize, setBannerSize] = React.useState(activity?.BannerSize || "16");
+    const [showBanner, setShowBanner] = React.useState(activity?.ShowBanner == 1 || false);
+
+    const { handleSubmit, control, reset } = useForm<BannerConfigValues>({ defaultValues });
+    const [color, setColor] = React.useState(white);
+    const updateColor = React.useCallback((evt: any, colorObj: IColor) => setColor(colorObj), []);
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                const request: any = data;
+                request.BannerColor = color.str;
+                request.BannerAction = request.BannerAction === true ? "1" : "0";
+                setBannerColor(request.BannerColor);
+                setBannerMessage(request.BannerContent);
+                setBannerSize(request.BannerSize);
+                setShowBanner(request.BannerAction == 1);
+                activity.setBanner(request);
+                closeForm();
+            },
+        )();
+    }, [activity, closeForm, color, handleSubmit]);
+
+    React.useEffect(() => {
+        if (!activity?.BannerColor) return;
+        setColor(getColorFromString(activity?.BannerColor));
+        const values = {
+            BannerAction: activity?.ShowBanner || 0,
+            BannerContent: activity?.BannerContent || "",
+            BannerScroll: activity?.BannerScroll || "",
+            BannerSize: activity?.BannerSize || ""
+        };
+        reset(values);
+    }, [activity?.BannerColor, activity?.BannerContent, activity?.BannerScroll, activity?.BannerSize, activity?.ShowBanner, reset]);
+
+    const BannerConfig = React.useMemo(() => () => {
+        return <MessageBox show={showForm} setShow={closeForm} title={nlsHPCC.SetBanner} minWidth={680}
+            footer={<>
+                <PrimaryButton text={nlsHPCC.OK} onClick={handleSubmit(onSubmit)} />
+                <DefaultButton text={nlsHPCC.Cancel} onClick={closeForm} />
+            </>}>
+            <Stack horizontal horizontalAlign="space-between">
+                <Stack.Item grow={2}>
+                    <Controller
+                        control={control} name="BannerAction"
+                        render={({
+                            field: { onChange, name: fieldName, value }
+                        }) => <Controller
+                                control={control} name={fieldName}
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value == 1} onChange={onChange} label={nlsHPCC.Enable} />}
+                            />
+                        }
+                    />
+                    <Controller
+                        control={control} name="BannerContent"
+                        render={({
+                            field: { onChange, name: fieldName, value },
+                            fieldState: { error }
+                        }) => <TextField
+                                name={fieldName}
+                                onChange={onChange}
+                                required={true}
+                                multiline
+                                autoAdjustHeight
+                                label={nlsHPCC.BannerMessage}
+                                value={value}
+                                errorMessage={error && error?.message}
+                            />}
+                        rules={{
+                            required: nlsHPCC.ValidationErrorRequired
+                        }}
+                    />
+                    <Controller
+                        control={control} name="BannerSize"
+                        render={({
+                            field: { onChange, name: fieldName, value },
+                            fieldState: { error }
+                        }) => <TextField
+                                name={fieldName}
+                                onChange={onChange}
+                                label={nlsHPCC.BannerSize}
+                                value={value}
+                                errorMessage={error && error?.message}
+                            />}
+                        rules={{
+                            pattern: {
+                                value: /^[0-9]+$/i,
+                                message: nlsHPCC.ValidationErrorEnterNumber
+                            }
+                        }}
+                    />
+                    <Controller
+                        control={control} name="BannerScroll"
+                        render={({
+                            field: { onChange, name: fieldName, value },
+                            fieldState: { error }
+                        }) => <TextField
+                                name={fieldName}
+                                onChange={onChange}
+                                label={nlsHPCC.BannerScroll}
+                                value={value}
+                                errorMessage={error && error?.message}
+                            />}
+                        rules={{
+                            pattern: {
+                                value: /^[0-9]+$/i,
+                                message: nlsHPCC.ValidationErrorEnterNumber
+                            }
+                        }}
+                    />
+                </Stack.Item>
+                <Stack.Item>
+                    <Label>{nlsHPCC.BannerColor}</Label>
+                    <ColorPicker
+                        onChange={updateColor}
+                        color={color}
+                    />
+                </Stack.Item>
+            </Stack>
+        </MessageBox>;
+    }, [closeForm, color, control, handleSubmit, onSubmit, showForm, updateColor]);
+
+    React.useEffect(() => {
+        setShowBanner(activity?.ShowBanner == 1);
+        setBannerMessage(activity?.BannerContent || "");
+        setBannerColor(activity?.BannerColor || "black");
+        setBannerSize(activity?.BannerSize || "16");
+    }, [activity?.BannerContent, activity?.ShowBanner, activity?.BannerColor, activity?.BannerSize]);
+
+    const BannerMessageBar = React.useMemo(() => () => {
+        return showBanner &&
+            <MessageBar
+                messageBarType={MessageBarType.warning}
+                onDismiss={() => setShowBanner(false)}
+                dismissButtonAriaLabel="Close"
+                isMultiline={false}
+                truncated={true}
+                overflowButtonAriaLabel="See More"
+                style={{
+                    color: bannerColor,
+                    fontSize: `${bannerSize}px`,
+                    lineHeight: `${bannerSize}px`
+                }}
+            >
+                {bannerMessage}
+            </MessageBar>
+            ;
+    }, [bannerColor, bannerMessage, bannerSize, showBanner]);
+
+    return [BannerMessageBar, BannerConfig];
+
+}

